Student list/IO-65

adbista opened the pull request:
feat: Enhance testing and functionality for user tags and student retrieval

This branch significantly enhances the application's testing suite and functionality concerning user tags and student data handling. Key updates include:

- Addition of tests for retrieving students using a combination of filters, sorting, and pagination. This enhances our ability to ensure the system behaves as expected under varied query conditions.
- Implementation of functionality for creating user tags, assigning these tags to users, and removing these connections, thereby improving data management capabilities.
- Modification of the User and UserTag models to include hashCode and equals methods along with JsonManagedReference and JsonBackReference annotations. These changes aim to improve object comparison and JSON serialization/deserialization processes.
- Introduction of UserTagDTO for better data transfer object management.
- Added student data retrieval, allowing for more sophisticated querying capabilities including sorting and filtering.

Commit:
[IO-23] add endpoint for retrieving students

Commit:
[IO-23] feat: add tags for users

Commit:
[IO-65] feat: Add tests for tag assignments and student retrieval with filters

Voitek2001 commented on src/main/java/edu/agh/dean/classesverifierbe/service/UserService.java:
```diff
@@ -12,12 +17,17 @@
     public User addUser(UserDTO userDTO) {
 
         if (userRepository.existsByIndexNumber(userDTO.getIndexNumber())) {
             throw new IllegalArgumentException("User with this index number already exists");
```
Add custom exceptions

Voitek2001 commented on src/main/java/edu/agh/dean/classesverifierbe/service/UserService.java:
```diff
@@ -40,6 +50,35 @@ public User toUser(UserDTO userDTO) {
+        User user = findUserByIndexNumber(index);
+        UserTag tag = findUserTagByName(tagName);
+        if (user.getUserTags().contains(tag)) {
+            throw new RuntimeException("User already has this tag");
```
as above

Voitek2001 reviewed (changes requested):
Seems nice but you should change exceptions e.g. RuntimeException for custom one

Commit:
[IO-65] feat: add custom Exceptions

Commit:
feat(IO-65): Enhance user filtering capabilities

- Support filtering by multiple tags in UserSpecifications.
- Introduce filtering by user status in UserService.

These enhancements allow for more refined querying of users based on their tags and current status, improving the application's flexibility in user management.
